### DISCUSSION TITLE
A little help understanding the state of the project

### DISCUSSION THREAD

**Discussion by brianredbeard** [terjeio/grblHAL#284] [General]
Howdy! First up, kudos on the forward progress with bringing grbl into the future.  Even just seeing more than five commits over a twelve month period is very exciting let alone the (comparative) break neck pace this project has.

I recently purchased one of @phil-barrett's grblHAL Teensy boards and was excited to dive in, get a firmware compiled, and start considering what I was going to tackle.  Assembly was trivial and I had no problems making some small tweaks (alternating the colors of some pin headers so i can differentiate them without putting reading glasses on, a matching 2x3 1.27mm pitch pin header for adding ethernet down the line, and [installing FRAM][fram] and additional SPI on the chip).

After looking through the various repositories (e.g. `terjeio/grblHAL`, `grblHAL/core`, `grblHAL/iMXRT1062`, etc), I'll admit it: 

I'm having a _really_ rough time of figuring out where effort is being focused, what needs help (be it development, testing, drivers, etc), if there is a "wish list", and most importantly _where no help is needed (or desired) because "drive by" contributions will be a distraction_.

As someone who makes lots of small contributions (mostly housekeeping type patches to help core contributors stay focused on the work which drives them) **I want to help but it's not clear where to start**.  

![italian spiderman](https://media.giphy.com/media/lTdaEMkLvHGzC/giphy.gif)

For example:

 - I can't seem to find any place where synchronous communication occurs.  Is there such a location? In lieu of that how is work organized?
 - Where is the canonical place for documentation?
   - The `README.md` in `grblHAL/core` is linking to a `changelog.md` in `terjeio/grblHAL` but is also referencing the wiki in `grblHAL/core`.  In `grblHAL/iMXRT1062` the directions point to https://github.com/terjeio/grblHAL/wiki/Compiling-GrblHAL while I also see that page is older than https://github.com/grblHAL/core/wiki/Compiling-GrblHAL (the latter was updated earlier today).
   - When I look at the `discussions` feature in `grblHAL/core` it's mostly empty and I recall reading a push to _discourage_ it's use until some critical mass occurs (which seems odd because trying to sync and de-duplicate content later is likely to be a pain in the ass).
   - When I look at the [HAL Reference][hal-reference] there's a datestamp on the bottom of the page in 2019 but that doesn't match the changelog of the page (I _think_ that may be due to migrating the wiki using a `git push` of the content, but I can't tell because there isn't any information wiki changelog outside of two entries which say "Updated Wiki" as the commit subject).
 - It seems like the present strategy is to rely on the native development environments provided by each MCU (e.g. CubeMX, the Arduino IDE, etc), is this the plan going forward? (_edit_: While about to submit this (I took my time over the course of ~1 hour to research and ensure I wasn't putting my foot in my mouth), I _**did**_ find this page https://github.com/grblHAL/core/wiki/Using-PlatformIO-for-Arduino-development.  Is this the direction to go in for the future?)
 - The git history seems to have been lost/purged as `terjeio/grblHAL` is fanning out into the various repositories (e.g. in `grblHAL/core` the commit message for the majority of the files is "test check in" and the rest of the files in the repository don't have a meaningful commit message through which to glean what's changing/getting fixed).  Are the new repositories just going to "start from zero" in terms of commit history?

I realize that the majority of these concerns are most accurately categorized as "nits" and I don't want to diminish the years of effort @terjeio has invested in resurrecting grbl.  This effort is Herculean and the outcome is impressive.

![quite impressed](https://media.giphy.com/media/JRaXdcSYrkr5rbrayI/giphy.gif)

[hal-reference]: https://github.com/grblHAL/core/wiki/Hardware-Abstraction-Layer-(HAL)----developer-reference
[fram]: https://www.grbl.org/single-post/eeprom-do-i-need-it

**Comment by brianredbeard** [comment]
Also, what is the license of grblHAL?

**Comment by phil-barrett** [comment]
There is no question that Terje's labors have yielded an amazing creation. I am continuously impressed by his willingness to help even the most clueless people.  I have nothing but admiration for him.

I think the license inherits from Grbl. (grbl/gnea?)  Definitely worth figuring out and explicitly stating.

There is a [road map](https://github.com/terjeio/grblHAL/wiki/Roadmap) in the Wiki though I suspect you have looked at it. Perhaps a more formalized approach is called for?

The documentation could use a reworking to pull it all together.  It has grown somewhat organically and is a bit spread out at this point. I think the wiki would make a good base for this but coalescing it in one place is far more important. 

There have been discussions of a "configurator" similar to what Marlin has.  This would make it easier for a non-developer to get a customized version.  One of the reasons GRBL was so successful is that a non-developer could set it up with very little hand holding. Unfortunately for grblHAL, supporting so many platforms makes this a much harder goal.

I would like to see some reorganizing of the project (under core?) to create a release process (along with version numbers).  Right now version numbers exist but aren't externalized very explicitly.

A general test suite and test plan would be a good addition to the project.

In terms of features, there are quite a few that would be worthwhile putting effort into.  Some work has been done on a few of these though there is a fair amount more needed for most. In no particular order:

- Odometer support, especially with ways to set maintenance triggers (like lubricate rail bearings every 10 KM or check motor brushes every 50 hours).  Would need some sort of UI as well.
- Pendant/MPG support.  This could be split into a number of different approaches.
-- Host USB connection to MPG
-- I2C connected control panel 
-- Encoder via digital inputs
- Modbus/RS485 driver 
- Torch Height Controller driver for a CNC plasma machine, PID testing/tuning
- Driver to support networked SD card (or other storage - QSPI devices for example)
- Control panel support for running GCode from local storage. There are a number of multicore processors that are emerging (iMXRT 1170 for example) that might be a very good fit. This, with the previous one would make for a really nice stand-alone CNC machine.

[Edit] I realized I used some incorrect terms here.  In grblHAL, a driver is the code that allows grblHAL to run on a specific microcontroller (I sometimes call that a port).  A plugIn is code that supports a specific feature, independent of the driver.  I meant PlugIn instead of driver in the list of features.

There are probably a lot features I have left off.

**Reply by brianredbeard** [reply]
This is IMMENSELY helpful.  I'm sure others will have thoughts / comments which would be good to capture.  

I think from my side I'm going to think about documentation a bit (since i'm at the intersection of _needing_ it as well as having experience working on exactly this sort of concern, e.g. [coreos/docs](https://github.com/coreos/docs).  There are plenty of lessons learned over working on that infrastructure for many years.  

For that specifically I'll make a proposal, give some time for feedback, and then pick up a shovel to start digging.

**Reply by phil-barrett** [reply]
That would be fantastic!

**Reply by phil-barrett** [reply]
One thing that should be part of the documentation is an expanded road map for the project.  This would clearly be a living document.

**Reply by Klippert** [reply]
I've been through quite a bit of a learning process myself over the last four months, and i'm happy to chip in with documentation. just point me in the right direction.

**Comment by Klippert** [comment]
That's a very good point. I guess one of the things i've noticed is that Grbl-hal developent is quite closely tied to io-sender development. but i think it would be fantastic if we could get some more structure on things, so it's easier to pitch in and know where to find things. I'm happy to help in any way i can.

**Comment by terjeio** [comment]
> I'm having a really rough time of figuring out where effort is being focused, what needs help (be it development, testing, drivers, etc), if there is a "wish list", and most importantly where no help is needed (or desired) because "drive by" contributions will be a distraction.

My wish is to get more contributors on board...

Better documentation is needed and I do not own the neccessary hardware to complete some of the plugins. So great if someone can contribute to those bits. IMO the core should be allowed to "mature" for a while, and perhaps most of the drivers too.

> I want to help but it's not clear where to start.

Neither am I as I do not understand github good enough... What I am sure about is that this repo should be phased out and activity moved to the [new one](https://github.com/grblHAL). I am thinking about adding [GitHub Pages](https://docs.github.com/en/pages) to the new one, would that be helpful? Opening up access to the new repo to a trusted core team would also be a good idea?

> It seems like the present strategy is to rely on the native development environments provided by each MCU (e.g. CubeMX, the Arduino IDE, etc), is this the plan going forward? (edit: While about to submit this (I took my time over the course of ~1 hour to research and ensure I wasn't putting my foot in my mouth), I did find this page https://github.com/grblHAL/core/wiki/Using-PlatformIO-for-Arduino-development. Is this the direction to go in for the future?)

I prefer to use the native environment and as much "bare metal" coding as possible. And I do not want to write another driver for a platform that is not debuggable - it is far too expensive. Adding scripts to compile the drivers for a variety of platforms should be possible, I see no need to standardize on a single one.

> The git history seems to have been lost/purged as terjeio/grblHAL is fanning out into the various repositories (e.g. in grblHAL/core the commit message for the majority of the files is "test check in" and the rest of the files in the repository don't have a meaningful commit message through which to glean what's changing/getting fixed). Are the new repositories just going to "start from zero" in terms of commit history?

Since there is no way (AFAICT) to migrate the commit history in a meaningful way it is going to start from zero. I do not think this is a big loss for users, and I can give developers read access to the master subversion repository if needed. The subversion repository has the commit history all the way back to the start in 2016 and has more commits/details.

>  I guess one of the things i've noticed is that Grbl-hal developent is quite closely tied to io-sender development.

That is true - and is a good thing? It has enabled me to test out and implement ideas that enhances grblHAL/sender interaction.

**Reply by Klippert** [reply]
I kind of like the closeness of iosender and grbl-hal - as @terjeio says, it really makes a lot of things easier.
It would seem that we just need to make a decision and set a date to move to the new repository, or am i missing something?
I'm happy to help out with documentation and other things, but if we're moving repo, i'm kind of waiting on hold. - And i get the notion that there are actually quite a few people willing to help, but also needing to be pointed in the right direction. 

if we can get everyone to run the same direction, i think we could accomplish something quite phenomenal...

**Reply by terjeio** [reply]
> It would seem that we just need to make a decision and set a date to move to the new repository, or am i missing something?

The move is complete for the souce code and the wiki, new issues, pull requests and discussions should be started there. The [readme](https://github.com/terjeio/grblHAL#grblhal) was updated earlier to inform about this change.

**Comment by phomann** [comment]
This all sounds very promising.
I'm also happy to help as I plan to use the teensy 4.1 platform for my openPnP pick and place machine conversion. From my perspective, the plugin architecture is a big plus. I'm still dealing with some of the rewiring required before I can start assisting with the development.
My current idea is to create a plugin to deal with some of the G and M codes not currently supported by GRBL but used by openPnP.

One thing I'd like to see that I don't think is there yet, is a compile time #define method for removing spindle control functionality, probe functionality, etc. I'd like be able to reassign these pins as general I/O.

As to a development environment, I agree that PlatformIO is probably the future. Marlin 2.0 uses it very effectively to manage different platforms and architectures.

Cheers,

Peter

**Reply by terjeio** [reply]
> One thing I'd like to see that I don't think is there yet, is a compile time #define method for removing spindle control functionality, probe functionality, etc. I'd like be able to reassign these pins as general I/O.

It might be a better idea to create an openPnP specific driver in order to avoid too much spagetthi code? The additonal G- and M-codes should still be implemented in a plugin to allow portability - perhaps on top of the ioport API:

https://github.com/terjeio/grblHAL/blob/4b95f3f7c8b82760aadd64f8957a29dad899fd54/drivers/IMXRT1062/grblHAL_Teensy4/src/grbl/hal.h#L92-L100

I have started work on extending this to support events (via interrupt callbacks).

> As to a development environment, I agree that PlatformIO is probably the future.

Nobody knows what the future will bring so I will not limit the development platform to a single one. Having the drivers as submodules allows one to chose whichever is best suited, and there is no need to wait for PlatformIO support for new MCUs.
